### PR TITLE
Packages: Change how required built-ins are polyfilled with Babel 7

### LINKED ISSUE #8727: Change how required built-ins are polyfilled
The current npm packages that being built for Gutenberg polyfill usage of some browser features that are not available on older browsers. This polyfill is global and affects any client who uses the gutenberg npm packages. This is unexpected and can alter the runtime environment in surprising ways. Additonally, some client environments do not require these polyfills and would work without them. Bundling them increases the overall bundle size and slows down loading.

Instead, gutenberg's packages should take one of two approaches: use use babel's `transform-runtime` plugin to replace problematic built-in usage, or stop polyfilling altogether and declare that host environments must polyfill certain features for Gutenberg to work on downlevel browsers. The later is the direction being taken by react-dom and some other libraries that require full ES5 compliance. 

Original discussion: https://github.com/WordPress/gutenberg/pull/8057#issuecomment-411415555

### LINKED ISSUE #9478: Packages that ship polyfills should depend on core-js@2
At the moment, many npm packages from Gutenberg ship with polyfills that use `core-js@2`. However, they only declare a dependency on `@babel/runtime-corejs2`. Since the built code has a bare require for `core-js`, the packages should also declare `core-js@2` as a dependency. Otherwise, version conflicts can occur, as the dependency is only transitive. If a client is using some other package that also relies on core-js, but another version, that version may end up being loaded instead of the version bundled with babel's runtime-corejs2.

For example:
```
npm i core-js@1
npm i @wordpress/i18n
```
Results in an npm tree with `core-js@1` in the root. When client code loads `@wordpress/i18n`, the bare require for `core-js` loads the version from the root and fails.

Ideally no polyfills would be shipping with Gutenberg packages, but that's #9171

cc @gziolo @jsnajdr

### PR DESCRIPTION
**Pull request by gziolo**
## Description
Fixes #8727.
Closes #9478.

Discussed during Core JS meeting: https://make.wordpress.org/core/2018/08/14/javascript-chat-summary-august-14/.

**Problem:** The current npm packages that being built for Gutenberg polyfill usage of some browser features that are not available on older browsers. This polyfill is global and affects any client who uses WordPress npm packages. This is unexpected and can alter the runtime environment in surprising ways.

**Action items:**
* [x] corejs flag should be set to false when generating distribution version for packages and set to 2 when bundling with Webpack.
* [x] we should inform developers consuming WordPress packages to install proper polyfills on their own.

This PR also updates Babel to the latest RC version since we need to update all occurrences of `@babel/runtime-corejs2` with `@babel/runtime` anyways.

~I also changed the configuration for the `build-module` distribution output. I took advantage of the new setting for `@babel/preset-env`: `"targets": { "esmodules": true }`~

~We use Babel twice to produce output for packages and then use Webpack to produce build files to consume inside WordPress, so we can take advantage of it and offer `build` version of package which works with all browsers and `build-module` version which works with all browsers that support `<script type="module"></script>`.~

## How has this been tested?
`npm run dev`
`npm run build`
`npm test`

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 

/cc @blowery @jsnajdr

### CODE REVIEW

**Comment by gziolo** [comment]
I still see imports like:
```js
require("core-js/modules/es6.array.find");
```

My bet is that `babel-preset-env` still need some tweaks.

**Comment by ntwb** [comment]
> The current npm packages that being built for Gutenberg polyfill usage of some browser features that are not available on older browsers.

Would this be due to the supported `@wordpress/browserslist-config` browsers matrix, which is included in `babel-preset-default`?
https://github.com/WordPress/gutenberg/blob/87780dffd9d9837ed0bf131c0a1585d9a6a20c85/packages/babel-preset-default/index.js#L9

**Comment by gziolo** [comment]
> Would this be due to the supported `@wordpress/browserslist-config` browsers matrix, which is included in `babel-preset-default`?

I don't think it's an issue, it's rather `useBuiltIns: 'usage',`.

**Comment by gziolo** [comment]
I was able to tweak Babel to ensure that we won't publish any references to `core-js` or `@babel/runtime-corejs2` polyfills to npm. I hope this aligns with what we discussed at the Core JS meeting last week.

**Comment by gziolo** [comment]
Do you happen to know any good example of the message that could be included to inform people using our packages to install required polyfills themselves?

It also raises the questions like:
- whether we should include polyfills for ES5 build? I assumed: no.
- which browsers should we target for `module` version? I assumed: only those which support  `<script type="module"></script>`.

**Comment by aduth** [comment]
> the message that could be included to inform people using our packages to install required polyfills themselves?

Part of this will require identifying what features are actually in use which could require need for a polyfill.

>which browsers should we target for module version?

It's tricky, since it's not only expected to be consumed by those browsers which support the script type, but also many transpilers like Webpack can use it for bundling from a dependency; in which case your package may include modern features inadvertently (because typically Babel won't be applied over an external dependency).

**Comment by youknowriad** [comment]
> In other words, we should leave the behavior as is and transpile all features but import & export and target the same browsers for both main and module. I will update PR accordingly.

FYI this is already what we do in master.

**Comment by gziolo** [comment]
> > the message that could be included to inform people using our packages to install required polyfills themselves?
> 
> Part of this will require identifying what features are actually in use which could require need for a polyfill.

I don't think this is an easy task to accomplish although it seems like a very expected feature to be included in `@babel/preset-env`. See related issue in Babel repository: https://github.com/babel/babel/issues/6625. I'm not quite sure what is the best way move forward here. I don't feel like it is an easy task to manually maintain which polyfills should be included for individual packages.

**Comment by gziolo** [comment]
Added 70d46ad which adds the following note to all README files of packages that need them:

> _This package assumes that your code will run in an ES5 environment. If you're using an environment that has limited or no support for ES5 such as lower versions of IE then using [@babel/polyfill](https://babeljs.io/docs/en/next/babel-polyfill) will add support for these methods. Learn more about it in [Babel docs](https://babeljs.io/docs/en/next/caveats)._

I also upgraded Babel to `7.0.0.-rc.2`.

**Comment by aduth** [comment]
When we say "assumes that your code will run in an ES5 environment", are we including things like `Promise`... or `Array#includes`? These are things I expect from `@babel/polyfill`, but they're also not ES5: They're ES2015+.

**Comment by gziolo** [comment]
> When we say "assumes that your code will run in an ES5 environment", are we including things like Promise... or Array#includes? These are things I expect from @babel/polyfill, but they're also not ES5: They're ES2015+.

I think I mixed up a few things :) I think this message should be closer to:

> _This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for ES2015+ such as lower versions of IE then using [core-js](https://github.com/zloirock/core-js) or [@babel/polyfill](https://babeljs.io/docs/en/next/babel-polyfill) will add support for these methods. Learn more about it in [Babel docs](https://babeljs.io/docs/en/next/caveats)._

**Comment by gziolo** [comment]
> The precise description of that enviroment is something like: only ES5 syntax is required, but the following ES2015+ builtins are expected to be present: [enumerated list of them].
>
> How to construct that enumerated list? Not easy. useBuiltins: 'usage' option of the env preset can help, but it's not reliable: until recently it didn't detect Promise.finally and there are many reported gaps like this one.

Unfortunately, it is almost impossible to create and what's even worse to maintain the list of polyfills for 30+ packages...

> By the way, useBuiltins: 'usage' is a noop after this PR: the env preset is configured not to include core-js polyfills at all, so usages detected by this feature go into /dev/null.

We transpile code twice, so `useBuiltins: 'usage'` is used with Webpack. You can double check in the top level `build` folder.

File: bin/packages/get-babel-config.js [hunk 1]
```diff
@@ -19,29 +19,55 @@ if ( ! process.env.SKIP_JSX_PRAGMA_TRANSFORM ) {
...
-						{},
-						preset.options,
-						{ modules: 'commonjs' }
-					) ];
-				}
-				return preset;
-			} ),
+			plugins: map(
+				plugins,
+				( plugin ) => overrideOptions( plugin, '@babel/plugin-transform-runtime', {
```
[the remark this answers is not in the document]
Comment (youknowriad) [reply bin/packages/get-babel-config.js]: Just read the docs https://babeljs.io/docs/en/next/babel-polyfill and https://babeljs.io/docs/en/next/babel-plugin-transform-runtime and I think it confirms my thoughts that we should just get rid of the runtime.

[the remark this answers is not in the document]
Comment (blowery) [reply bin/packages/get-babel-config.js]: We should be using `useBuiltIns: false`, not `true` and using the transform runtime. We should not be shipping polyfills that modify global prototypes.

[the remark this answers is not in the document]
Comment (youknowriad) [reply bin/packages/get-babel-config.js]: @blowery We are using `useBuiltIns: false` in the packages (npm) but I was referring to `useBuiltIns: true` used for Gutenberg.

[the remark this answers is not in the document]
Comment (youknowriad) [reply bin/packages/get-babel-config.js]: This raises the question of whether we should keep the runtime transform for the packages. I personally don't see any difference between those polyfills and the remaining ones not included in the runtime (aside from the way they are included), which means I think we should remove all the polyfills and not only the ones added by `babel-preset-env`

[the remark this answers is not in the document]
Comment (gziolo) [reply bin/packages/get-babel-config.js]: I'm not quite sure. I followed the configuration from Calypso which was reviewed by @hzoo 😄 

It has both `@babel/preset-env` with `useBuiltIns: "entry"` (https://github.com/Automattic/wp-calypso/blob/master/babel.config.js#L22) and `@babel/plugin-transform-runtime` with `corejs: false` (https://github.com/Automattic/wp-calypso/blob/master/babel.config.js#L51).

[the remark this answers is not in the document]
Comment (jsnajdr) [reply bin/packages/get-babel-config.js]: Yes, there are Babel helpers that are not polyfills. When transpiling classes, for example, Babel uses helper functions with names like `_classCallCheck`, `_possibleConstructorReturn` or `_inherits`. The runtime transform plugin prevents them from being duplicated all the time and imports them from `@babel/runtime/helpers` instead.

Another thing that is not a CoreJS polyfill is the regenerator runtime used when transpiling generators and async/await.

The CoreJS and regenerator are "polyfills" in a sense that they might be provided by the platform already. Babel helpers never are though.

[the remark this answers is not in the document]
Comment (youknowriad) [reply bin/packages/get-babel-config.js]: Thanks for the explanation @jsnajdr it makes sense 👍

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Comment by aduth** [comment]
I can confirm this appears to have broken the editor in IE11 due to the lack of a provided polyfill, verified via the following (_very-naive-definitely-not-final_) fix:

```diff
diff --git a/lib/client-assets.php b/lib/client-assets.php
index 7ad3546c5..d2bcabd21 100644
--- a/lib/client-assets.php
+++ b/lib/client-assets.php
@@ -66,6 +66,11 @@ function gutenberg_get_script_polyfill( $tests ) {
 	return $polyfill;
 }
 
+function gutenberg_ie11_ಥ_ಥ() {
+	?><script src="https://cdnjs.cloudflare.com/ajax/libs/babel-polyfill/7.0.0/polyfill.min.js"></script><?php
+}
+add_action( 'admin_enqueue_scripts', 'gutenberg_ie11_ಥ_ಥ', 0 );
+
 /**
  * Registers the main TinyMCE scripts.
  */
```

In trying to come to a more solid solution, I'm left wondering what precisely we're expecting people to polyfill. Is it ES2015? How about things like `Array#includes` (ES2016) or `Object.values` (ES2017)?

**Comment by aduth** [comment]
Tracking IE11 issue for release candidate: #9640

**Comment by gziolo** [comment]
> Did we verify this assumption for Gutenberg itself? 😄 There appear to be some issues loading the editor in current master in IE11 which on preliminary testing seem to be attributable to changes here.

We should 🤦‍♂️ I assumed this is going to be transparent for Gutenberg itself...

**Comment by aduth** [comment]
>In trying to come to a more solid solution, I'm left wondering what precisely we're expecting people to polyfill. Is it ES2015? How about things like `Array#includes` (ES2016) or `Object.values` (ES2017)?

To expand on this with an answer to my own question:

Babel's polyfill intends to provide anything which is part of a published ECMAScript specification v2015 or newer.

>This will emulate a full ES2015+ environment (no < Stage 4 proposals) 

https://babeljs.io/docs/en/babel-polyfill.html

This is the reason why, in #9643, I chose the name `wp-polyfill-ecmascript` as the script handle, as it is a forever-moving target intended to follow the latest ECMAScript specification.

**Comment by aduth** [comment]
Well, to be fair, if one assumes that `babel-polyfill` is a polyfill for ECMAScript, which is how I've taken it to mean, then it's entirely expected that it should not include `fetch`, as it's not part of any ECMAScript specification. 

https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-262.pdf

Rather, it's a WHATWG (and I assume W3) specification:

https://fetch.spec.whatwg.org/

**Pull request merged**